### DISCUSSION TITLE
ipcRenderer  in imodel desktop-starter project

### DISCUSSION THREAD

**Discussion by paprophet** [iTwin/itwinjs-core#2086] [Q&A] [answered]
i want define a button and onclik the button (in front ) and the electron main process (src/backend/main.ts ) can console a string

**Comment by kabentley** [comment, marked as answer]
If you want to create your own Ipc interface, see [this article](https://www.itwinjs.org/learning/ipcinterface/#creating-your-own-ipc-interfaces-and-ipchandlers).

Here's an example: [interface](https://github.com/imodeljs/imodeljs/blob/13b8f1920c15b654989cdd49fd3ebb45cd7d22b4/test-apps/display-test-app/src/common/DtaIpcInterface.ts#L8-L10), [registration](https://github.com/imodeljs/imodeljs/blob/13b8f1920c15b654989cdd49fd3ebb45cd7d22b4/test-apps/display-test-app/src/backend/DtaElectronMain.ts#L50), [backend](https://github.com/imodeljs/imodeljs/blob/13b8f1920c15b654989cdd49fd3ebb45cd7d22b4/test-apps/display-test-app/src/backend/DtaElectronMain.ts#L33), [frontend](https://github.com/imodeljs/imodeljs/blob/13b8f1920c15b654989cdd49fd3ebb45cd7d22b4/test-apps/display-test-app/src/frontend/App.ts#L78-L82)

**Reply by paprophet** [reply]
thank you !

**Reply by paprophet** [reply]
i want download  file and show the progress of the download  on the front  by electron app.

**Reply by kabentley** [reply]
see [`NativeApp.requestDownloadBriefcase`](https://github.com/imodeljs/imodeljs/blob/54a631a35d21fa7a18fc126cc0934ac4044116d7/core/frontend/src/NativeApp.ts#L218). It is currently marked @internal, but will become public in 3.0.

Example tests [here](https://github.com/imodeljs/imodeljs/blob/master/full-stack-tests/native-app/src/frontend/hub/Download.test.ts).

**Reply by paprophet** [reply]
when use the electron in front 
import { ipcRenderer } from "electron";
show  error 

![1231231](https://user-images.githubusercontent.com/18022720/131210127-53ffd5c6-b0a5-45b6-afc7-78d7c1ecc8dc.png)

**Reply by kabentley** [reply]
Please show your code. Why are you importing ipcRenderer?
